File: engines/python/inserter_mysql.py
import csv
import mysql.connector


def _connect(conn_params: dict):
    # use_pure=True + charset=utf8mb4: на Windows с русской локалью
    # mysql.connector без этих опций может ронять процесс на декоде ошибок.
    conn = mysql.connector.connect(
        host=conn_params["host"],
        port=conn_params["port"],
        user=conn_params["user"],
        password=conn_params["password"],
        database=conn_params["database"],
        allow_local_infile=True,
        charset="utf8mb4",
        use_pure=True,
    )
    return conn


def _quote(name: str) -> str:
    clean = name.strip().strip("`").replace("`", "``")
    return f"`{clean}`"


def _detect_lines_terminator(path: str) -> str:
    """Возвращает строку-аргумент для LOAD DATA LINES TERMINATED BY.
    Без декодирования файла: читаем первые 8 КБ, ищем CRLF — типичен для
    CSV, созданных на Windows (Excel, Python csv с дефолтным newline и т.п.).
    MySQL LOAD DATA INFILE строгий: при несовпадении терминатора всё
    содержимое воспринимается как одна «строка», и IGNORE 1 ROWS оставляет
    1 «битую» запись."""
    with open(path, "rb") as f:
        chunk = f.read(8192)
    if b"\r\n" in chunk:
        return r"\r\n"  # 4 символа: \, r, \, n — MySQL раскроет как CRLF
    return r"\n"


def count_rows(conn_params: dict, table_name: str) -> int:
    """Возвращает фактическое число строк в таблице из самой БД.
    Используется в main.py для независимой проверки результата вставки."""
    conn = _connect(conn_params)
    cursor = conn.cursor()
    try:
        cursor.execute(f"SELECT COUNT(*) FROM {_quote(table_name)}")
        row = cursor.fetchone()
        return int(row[0]) if row else 0
    finally:
        cursor.close()
        conn.close()


def default_insert(conn_params: dict, csv_file: str, table_name: str) -> int:
    conn = _connect(conn_params)
    cursor = conn.cursor()
    count = 0
    try:
        with open(csv_file, "r", newline="", encoding="utf-8") as f:
            for row in csv.DictReader(f):
                col_names = ", ".join(_quote(c) for c in row.keys())
                placeholders = ", ".join(["%s"] * len(row))
                cursor.execute(
                    f"INSERT INTO {_quote(table_name)} ({col_names}) VALUES ({placeholders})",
                    list(row.values()),
                )
                count += 1
        conn.commit()
        return count
    except Exception:
        conn.rollback()
        raise
    finally:
        cursor.close()
        conn.close()


def bulk_insert(
    conn_params: dict, csv_file: str, table_name: str, batch_size: int = 1000
) -> int:
    conn = _connect(conn_params)
    cursor = conn.cursor()
    count = 0
    try:
        with open(csv_file, "r", newline="", encoding="utf-8") as f:
            reader = csv.DictReader(f)
            col_names = ", ".join(_quote(c) for c in reader.fieldnames)
            placeholders = ", ".join(["%s"] * len(reader.fieldnames))
            sql = f"INSERT INTO {_quote(table_name)} ({col_names}) VALUES ({placeholders})"

            batch = []
            for row in reader:
                batch.append(list(row.values()))
                if len(batch) >= batch_size:
                    cursor.executemany(sql, batch)
                    count += len(batch)
                    batch.clear()
            if batch:
                cursor.executemany(sql, batch)
                count += len(batch)

        conn.commit()
        return count
    except Exception:
        conn.rollback()
        raise
    finally:
        cursor.close()
        conn.close()


def file_insert(conn_params: dict, csv_file: str, table_name: str) -> int:
    conn = _connect(conn_params)
    cursor = conn.cursor()
    try:
        with open(csv_file, "r", newline="", encoding="utf-8") as f:
            row_count = sum(1 for _ in csv.DictReader(f))

        abs_path = csv_file.replace("\\", "/")
        line_term = _detect_lines_terminator(csv_file)
        cursor.execute(
            f"""
            LOAD DATA LOCAL INFILE '{abs_path}'
            INTO TABLE {_quote(table_name)}
            FIELDS TERMINATED BY ','
            OPTIONALLY ENCLOSED BY '"'
            LINES TERMINATED BY '{line_term}'
            IGNORE 1 ROWS
        """
        )
        conn.commit()
        return row_count
    except Exception:
        conn.rollback()
        raise
    finally:
        cursor.close()
        conn.close()

File: engines/rust/src/inserter_mysql.rs
use crate::csv::{CsvStream, clean_identifier};
use crate::inserter::{ConnParams, Inserter};
use anyhow::{Result, anyhow};
use mysql::prelude::*;
use mysql::*;
use std::io::Read;

/// Определяет line-ending CSV-файла без декодирования.
/// Возвращает строку-аргумент для LOAD DATA LINES TERMINATED BY:
/// `\r\n` (Windows-стиль) или `\n`. Без этого LOAD DATA на Windows-файле
/// может потерять разбиение строк и оставить в таблице 1 битую запись.
fn detect_lines_terminator(path: &str) -> std::io::Result<&'static str> {
    let mut file = std::fs::File::open(path)?;
    let mut buf = [0u8; 8192];
    let n = file.read(&mut buf)?;
    if buf[..n].windows(2).any(|w| w == b"\r\n") {
        Ok(r"\r\n")
    } else {
        Ok(r"\n")
    }
}

pub struct MySQLInserter {
    conn: Conn,
}

impl MySQLInserter {
    pub fn new(p: &ConnParams) -> Result<Self> {
        let url = format!(
            "mysql://{}:{}@{}:{}/{}",
            p.user, p.password, p.host, p.port, p.database
        );

        let builder = OptsBuilder::from_opts(Opts::from_url(&url)?)
            .local_infile_handler(Some(LocalInfileHandler::new(
                |file_name: &[u8], infile: &mut mysql::LocalInfile<'_>| {
                    // Стримим файл прямо в сокет, без загрузки в Vec<u8>.
                    let path = String::from_utf8_lossy(file_name).to_string();
                    let mut file = std::fs::File::open(&path)?;
                    std::io::copy(&mut file, infile)?;
                    Ok(())
                },
            )));

        let conn = Conn::new(builder)?;
        Ok(Self { conn })
    }

    fn quote(name: &str) -> String {
        let clean = clean_identifier(name);
        format!("`{}`", clean.replace('`', "``"))
    }

    fn build_cols(headers: &[String]) -> String {
        headers.iter()
            .map(|h| Self::quote(h))
            .collect::<Vec<_>>()
            .join(", ")
    }
}

impl Inserter for MySQLInserter {

    fn count_rows(&mut self, table: &str) -> Result<usize> {
        let sql = format!("SELECT COUNT(*) FROM {}", Self::quote(table));
        let row: Option<i64> = self.conn.query_first(&sql)?;
        Ok(row.unwrap_or(0) as usize)
    }

    fn default_insert(&mut self, csv_file: &str, table: &str) -> Result<usize> {
        let mut stream = CsvStream::open(csv_file)?;
        let cols  = Self::build_cols(&stream.headers);
        let phs   = vec!["?"; stream.headers.len()].join(", ");
        let sql   = format!("INSERT INTO {} ({}) VALUES ({})",
                            Self::quote(table), cols, phs);

        let mut tx = self.conn.start_transaction(TxOpts::default())?;
        let mut count = 0;
        while let Some(row) = stream.next() {
            let row = row?;
            let params: Vec<Value> = row.iter()
                .map(|v| Value::Bytes(v.as_bytes().to_vec()))
                .collect();
            tx.exec_drop(&sql, params)?;
            count += 1;
        }
        tx.commit()?;
        Ok(count)
    }

    fn bulk_insert(&mut self, csv_file: &str, table: &str,
                   batch_size: usize) -> Result<usize> {
        let mut stream = CsvStream::open(csv_file)?;
        let cols   = Self::build_cols(&stream.headers);
        let qtable = Self::quote(table);
        let ncols  = stream.headers.len();
        let mut total = 0;

        // Шаблон одной группы плейсхолдеров `(?, ?, …)` строим один раз.
        let row_phs = format!("({})", vec!["?"; ncols].join(", "));

        let mut batch: Vec<Vec<String>> = Vec::with_capacity(batch_size);
        let mut eof = false;

        while !eof {
            match stream.next() {
                Some(Ok(row)) => batch.push(row),
                Some(Err(e))  => return Err(e),
                None          => eof = true,
            }

            // Шлём пачку, когда она накопилась до batch_size или CSV кончился.
            if (batch.len() >= batch_size || (eof && !batch.is_empty())) && !batch.is_empty() {
                let all_phs = vec![row_phs.as_str(); batch.len()].join(", ");
                let sql = format!("INSERT INTO {} ({}) VALUES {}",
                                  qtable, cols, all_phs);

                let params: Vec<Value> = batch.iter()
                    .flat_map(|row| row.iter()
                        .map(|v| Value::Bytes(v.as_bytes().to_vec())))
                    .collect();

                let mut tx = self.conn.start_transaction(TxOpts::default())?;
                tx.exec_drop(&sql, params)?;
                tx.commit()?;
                total += batch.len();
                batch.clear();
            }
        }

        Ok(total)
    }

    fn file_insert(&mut self, csv_file: &str, table: &str) -> Result<usize> {
        // Парсить файл не нужно — LOAD DATA INFILE даёт серверу прочитать его
        // самостоятельно. Возвращаемое значение метода больше не используется
        // в main: фактическое число строк main получает через count_rows()
        // уже после замера времени.
        let raw_path = std::path::PathBuf::from(csv_file);
        let abs_path = if raw_path.is_absolute() {
            raw_path
        } else {
            std::env::current_dir()?.join(raw_path)
        };

        // Backslash → forward slash: безопасно и для Windows API, и для MySQL.
        let sql_path = abs_path.to_string_lossy().replace('\\', "/");

        let line_term = detect_lines_terminator(csv_file)
            .map_err(|e| anyhow!("detect line endings: {}", e))?;

        let sql = format!(
            "LOAD DATA LOCAL INFILE '{}' \
            INTO TABLE {} \
            FIELDS TERMINATED BY ',' \
            OPTIONALLY ENCLOSED BY '\"' \
            LINES TERMINATED BY '{}' \
            IGNORE 1 ROWS",
            sql_path,
            Self::quote(table),
            line_term
        );

        self.conn
            .query_drop(sql)
            .map_err(|e| anyhow!("LOAD DATA error: {}", e))?;

        Ok(0)
    }
}
